cluster/images/etcd/Makefile

```diff
@@ -114,5 +116,124 @@ ifeq ($(ARCH),amd64)
+			'[ $$(cat /var/etcd/data/version.txt) = $(ETCD2_ROLLBACK_NEW_TAG)/etcd3 ]'
+
+	@echo "Rolling back to the previous version of etcd and recording keyspace to a flat file."
+	docker run --tty --interactive -v $(TEMP_DIR)/rollback-etcd2:/var/etcd \
+		-e "TARGET_STORAGE=etcd2" \
+		-e "TARGET_VERSION=$(ETCD2_ROLLBACK_OLD_TAG)" \
+		-e "DATA_DIRECTORY=/var/etcd/data" \
+		gcr.io/google_containers/etcd-$(ARCH):$(REGISTRY_TAG) /bin/sh -c \
+			'/usr/local/bin/migrate-if-needed.sh && \
+			source /usr/local/bin/start-stop-etcd.sh && \
+			START_STORAGE=etcd2 START_VERSION=$(ETCD2_ROLLBACK_OLD_TAG) start_etcd && \
+			/usr/local/bin/etcdctl-$(ETCD2_ROLLBACK_OLD_TAG) --endpoint 127.0.0.1:$${ETCD_PORT} get /registry/k1 > /var/etcd/keyspace.txt && \
+			stop_etcd'
+
+	@echo "Checking if rollback successfully downgraded etcd to $(ETCD2_ROLLBACK_OLD_TAG)"
+	docker run --tty --interactive -v $(TEMP_DIR)/rollback-etcd2:/var/etcd \
+		gcr.io/google_containers/etcd-$(ARCH):$(REGISTRY_TAG) /bin/sh -c \
+			'[ $$(cat /var/etcd/data/version.txt) = $(ETCD2_ROLLBACK_OLD_TAG)/etcd2 ] && \
+			 grep -q value1 /var/etcd/keyspace.txt'
+
+# Test a rollback from the latest version to the previous version.
+test-rollback:
+	mkdir -p $(TEMP_DIR)/rollback-test
+	cd $(TEMP_DIR)/rollback-test
+
+	@echo "Starting $(REGISTRY_TAG) etcd and writing some sample data."
+	docker run --tty --interactive -v $(TEMP_DIR)/rollback-test:/var/etcd \
+		-e "TARGET_STORAGE=etcd3" \
+		-e "TARGET_VERSION=$(REGISTRY_TAG)" \
+		-e "DATA_DIRECTORY=/var/etcd/data" \
+		gcr.io/google_containers/etcd-$(ARCH):$(REGISTRY_TAG) /bin/sh -c \
+			'/usr/local/bin/migrate-if-needed.sh && \
+			source /usr/local/bin/start-stop-etcd.sh && \
+			START_STORAGE=etcd3 START_VERSION=$(REGISTRY_TAG) start_etcd && \
+			ETCDCTL_API=3 /usr/local/bin/etcdctl --endpoints http://127.0.0.1:$${ETCD_PORT} put /registry/k1 value1 && \
+			stop_etcd'
+
+	@echo "Rolling back to the previous version of etcd and recording keyspace to a flat file."
+	docker run --tty --interactive -v $(TEMP_DIR)/rollback-test:/var/etcd \
+		-e "TARGET_STORAGE=etcd3" \
+		-e "TARGET_VERSION=$(ROLLBACK_REGISTRY_TAG)" \
+		-e "DATA_DIRECTORY=/var/etcd/data" \
+		gcr.io/google_containers/etcd-$(ARCH):$(REGISTRY_TAG) /bin/sh -c \
+			'/usr/local/bin/migrate-if-needed.sh && \
+			source /usr/local/bin/start-stop-etcd.sh && \
+			START_STORAGE=etcd3 START_VERSION=$(ROLLBACK_REGISTRY_TAG) start_etcd && \
+			ETCDCTL_API=3 /usr/local/bin/etcdctl --endpoints http://127.0.0.1:$${ETCD_PORT} get --prefix / > /var/etcd/keyspace.txt && \
+			stop_etcd'
+
+	@echo "Checking if rollback successfully downgraded etcd to $(ROLLBACK_REGISTRY_TAG)"
+	docker run --tty --interactive -v $(TEMP_DIR)/rollback-test:/var/etcd \
+		gcr.io/google_containers/etcd-$(ARCH):$(REGISTRY_TAG) /bin/sh -c \
+			'[ $$(cat /var/etcd/data/version.txt) = $(ROLLBACK_REGISTRY_TAG)/etcd3 ] && \
+			 grep -q value1 /var/etcd/keyspace.txt'
```

Comment:
I looked into it briefly and the main reason I kept them separate is because we will be deleting the etcd support later this year, and by keeping the tests separate we make the deletion trivial.